### ISSUE TITLE
Expression recursion level exceeded not detected

### ISSUE THREAD

**Issue by Ubiquitous** [koalaman/shellcheck#852]
#### For bugs
- Rule Id (if any, e.g. SC1000): 
- My shellcheck version (`shellcheck --version` or "online"): 
- [X] I tried on shellcheck.net and verified that this is still a problem on the latest commit
- [ ] It's not reproducible on shellcheck.net, but I think that's because it's an OS, configuration or encoding issue

#### For new checks and feature suggestions
- [X] shellcheck.net (i.e. the latest commit) currently gives no useful warnings about this
- [X] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

#### Here's a snippet or screenshot that shows the problem:

```
#!/bin/bash

n="n"
y="y"
while [[ $y -ne $n ]]
do
    ...
done
```

#### Here's what shellcheck currently says:
```
$ shellcheck myscript
No issues detected!
```

#### Here's what I wanted or expected to see:
```
line 5: [[: y: expression recursion level exceeded (error token is "y")
```

I might suggest checking whether or not a variable is actually an integer when using comparison operators such as `-ne`, `-gt`, etc. otherwise it might possibly lead to someone thinking that `-ne` is fine for comparing string values.

**Comment by nihilus** [comment]
The major problem here is that shellscripts are not statically typed; so the comparsion when used on variables cannot easily be deduced if it is done on a string or an integer (or a float).

**Comment by kurahaupo** [comment]
On Sun, 19 Mar 2017, Ubiquitous wrote:
> #### Here's what I wanted or expected to see:
> ```
> line 5: [[: y: expression recursion level exceeded (error token is "y")
> ```
>
> I might suggest checking whether or not a variable is actually an integer when using comparison operators such as `-ne`, `-gt`, etc. otherwise it might possibly lead to someone thinking that `-ne` is fine for comparing string values.

It's worth remembering that Shellcheck is a static analysis tool, not a
shell execution environment

I suggest reading up on "The Halting Problem".


This means there are cases where even in theory it cannot be determined
whether a self-reference will exceed the stack limit; for example, consider

  factorial='n>1?n--*factorial:1'

Whether or not that blows up depends on the initial value of n.
If you think that it suffices to have n statically defined, consider
instead

  ex='(n/=n&-n)==1 ? 0 : (n+=n*2+1, 1+ex)'

which, afaik, has has no proof that it must terminate; there's no
short-circuit -- the only way to tell how long it will take to run is to run
it, and if it's "forever" then you'll never know.

(As an aside, for the life of me I can't see why people insist on using
[[...]] for doing math, when that's what ((...)) is for.)

**Comment by koalaman** [comment]
Obviously you're right, but it's ShellCheck's decision not to try to figure out the shenanigans that indirect access can cause (due to the rabbit hole reasons mentioned above).
